src/Shared/TaskParameter.cs

```diff
@@ -537,6 +583,100 @@ private bool TranslateNullable<T>(ITranslator translator, T value)
             return haveRef;
         }
 
+        /// <summary>
+        /// Serializes or deserializes the value type instance wrapped by this <see cref="TaskParameter"/>.
+        /// </summary>
+        /// <remarks>
+        /// The value type is converted to/from string using the <see cref="Convert"/> class. Note that we require
+        /// task parameter types to be <see cref="IConvertible"/> so this conversion is guaranteed to work for parameters
+        /// that have made it this far.
+        /// </remarks>
+        private void TranslateValueType(ITranslator translator)
+        {
+            string typeName = null;
+            string valueString = null;
+
+            if (translator.Mode == TranslationDirection.WriteToStream)
+            {
+                Type type = _wrappedParameter.GetType();
+
+                // Don't allow non-serializable types to be translated to keep the same limitation as the previous
+                // BinaryFormatter-based implementation.
+#pragma warning disable SYSLIB0050
+                if (!type.GetTypeInfo().IsSerializable)
+#pragma warning restore SYSLIB0050
+                {
+                    throw new NotSupportedException($"{type} cannot be used as a task parameter type because it is not serializable.");
+                }
+
+                typeName = type.AssemblyQualifiedName;
+                valueString = Convert.ToString(_wrappedParameter, CultureInfo.InvariantCulture);
+            }
+
+            translator.Translate(ref typeName);
+            translator.Translate(ref valueString);
+
+            if (translator.Mode == TranslationDirection.ReadFromStream)
+            {
+                Type type = Type.GetType(typeName);
+                _wrappedParameter = Convert.ChangeType(valueString, type, CultureInfo.InvariantCulture);
+            }
+        }
+
+        /// <summary>
+        /// Serializes or deserializes the value type array instance wrapped by this <see cref="TaskParameter"/>.
+        /// </summary>
+        /// <remarks>
+        /// The array is assumed to be non-null.
+        /// </remarks>
+        private void TranslateValueTypeArray(ITranslator translator)
+        {
+            if (translator.Mode == TranslationDirection.WriteToStream)
+            {
+                Array array = (Array)_wrappedParameter;
+                Type arrayType = array.GetType();
+
+                // Don't allow non-serializable types to be translated to keep the same limitation as the previous
+                // BinaryFormatter-based implementation.
+#pragma warning disable SYSLIB0050
+                if (!arrayType.GetTypeInfo().IsSerializable)
```

Comment:
Since our long-term plan should be to kill this functionality, types are still required to be serializable so that we are as strict as the current logic, i.e. we don't want to open it up to more types. You're right that this limitation currently exists only for tasks hosted in the task host and while it feels artificial for the new code, I slightly prefer having this.

Reply:
I have updated the PR and we no longer have to discontinue this functionality. Consequently I have removed the restriction. Thank you!